File: sdk/cpp/tml_client.cpp
/**
 * TML C++ Client Implementation - Main client implementation for TML SDK
 * 
 * Path: /sdk/cpp/tml_client.cpp
 * Version: 1.0.0
 * Creator: Lev Goukassian (ORCID: 0009-0006-5966-1243)
 * 
 * This implements the TML client interface for Sacred Zero enforcement,
 * Always Memory logging, environmental monitoring, and multi-stakeholder
 * governance with full ecosystem consideration.
 */

#include "tml_client.h"
#include "sacred_zero.h"
#include "always_memory.h"
#include <mutex>
#include <thread>
#include <queue>
#include <condition_variable>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <random>
#include <cmath>
#include <algorithm>
#include <curl/curl.h>
#include <json/json.h>

namespace TML {

// ========== Internal Implementation Class ==========

class TMLClient::Impl {
public:
    // Configuration
    std::shared_ptr<Config> config;
    
    // Core components
    std::unique_ptr<SacredZeroTrigger> sacred_zero;
    std::unique_ptr<AlwaysMemoryLogger> always_memory;
    std::unique_ptr<GuardianConnection> guardian;
    
    // State management
    std::atomic<bool> connected{false};
    std::atomic<bool> audit_mode{false};
    mutable std::mutex state_mutex;
    
    // Error handling
    std::shared_ptr<TMLError> last_error;
    std::function<void(const TMLError&)> error_callback;
    mutable std::mutex error_mutex;
    
    // Sacred Zero callbacks
    std::vector<std::function<void(const SacredZeroResult&)>> sacred_zero_callbacks;
    std::mutex sacred_zero_mutex;
    
    // Event subscriptions
    struct Subscription {
        std::vector<std::string> event_types;
        std::function<void(const Metadata&)> callback;
        bool active;
    };
    std::map<std::string, Subscription> subscriptions;
    std::mutex subscription_mutex;
    std::thread event_thread;
    std::atomic<bool> event_thread_running{false};
    
    // Metrics
    std::map<std::string, double> metrics;
    std::mutex metrics_mutex;
    
    // Statistics
    struct Statistics {
        uint64_t sacred_zero_evaluations{0};
        uint64_t sacred_zero_triggers{0};
        uint64_t logs_sent{0};
        uint64_t environmental_checks{0};
        double total_carbon_kg{0.0};
        double total_water_liters{0.0};
    } stats;
    
    // Constructor
    Impl(std::shared_ptr<Config> cfg) : config(cfg) {
        if (!config) {
            config = std::make_shared<Config>(GetDefaultConfig());
        }
        InitializeComponents();
    }
    
    // Destructor
    ~Impl() {
        if (event_thread_running) {
            event_thread_running = false;
            if (event_thread.joinable()) {
                event_thread.join();
            }
        }
        if (connected) {
            Disconnect();
        }
    }
    
    // Initialize core components
    void InitializeComponents() {
        sacred_zero = std::make_unique<SacredZeroTrigger>(config);
        always_memory = std::make_unique<AlwaysMemoryLogger>(config);
        // Guardian connection would be initialized here in production
    }
    
    // Default configuration
    static Config GetDefaultConfig() {
        Config cfg;
        cfg.guardian_url = "https://guardian.tml-network.org";
        cfg.environment = Environment::PRODUCTION;
        cfg.block_on_sacred_zero = true;
        cfg.carbon_threshold_kg = 1000.0;
        cfg.water_threshold_liters = 10000.0;
        cfg.discrimination_threshold = 0.2;
        cfg.require_tee = false;
        cfg.audit_mode_enabled = false;
        return cfg;
    }
    
    // Generate unique ID
    static std::string GenerateID(const std::string& prefix) {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        static std::uniform_int_distribution<> dis(0, 15);
        
        std::stringstream ss;
        ss << prefix << "-";
        ss << std::time(nullptr) << "-";
        
        for (int i = 0; i < 8; i++) {
            ss << std::hex << dis(gen);
        }
        
        return ss.str();
    }
    
    // Calculate carbon impact
    double CalculateCarbonImpact(const ResourceUsage& resources) {
        // Simplified calculation - in production this would use detailed models
        double power_kw = (resources.cpu_cores * 0.05) + 
                         (resources.gpu_count * 0.3) +
                         (resources.memory_gb * 0.003);
        
        double hours = std::chrono::duration_cast<std::chrono::hours>(resources.duration).count();
        double energy_kwh = power_kw * hours;
        
        return energy_kwh * resources.carbon_intensity;
    }
    
    // Calculate water impact
    double CalculateWaterImpact(const ResourceUsage& resources) {
        // Data center cooling water usage estimation
        double hours = std::chrono::duration_cast<std::chrono::hours>(resources.duration).count();
        double base_water = 1.8; // Liters per kWh (typical data center)
        
        double power_kw = (resources.cpu_cores * 0.05) + 
                         (resources.gpu_count * 0.3);
        
        return power_kw * hours * base_water;
    }
    
    // Process event stream
    void ProcessEventStream() {
        while (event_thread_running) {
            // In production, this would receive events from Guardian
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            
            // Check for subscriptions and deliver events
            std::lock_guard<std::mutex> lock(subscription_mutex);
            for (auto& [id, sub] : subscriptions) {
                if (sub.active) {
                    // Simulate event delivery
                    // In production, real events would be delivered here
                }
            }
        }
    }
};

// ========== TMLClient Implementation ==========

TMLClient::TMLClient(std::shared_ptr<Config> config) 
    : pImpl(std::make_unique<Impl>(config)) {
}

TMLClient::~TMLClient() = default;

TMLClient::TMLClient(TMLClient&&) noexcept = default;
TMLClient& TMLClient::operator=(TMLClient&&) noexcept = default;

// ========== Connection Management ==========

bool TMLClient::Connect() {
    std::lock_guard<std::mutex> lock(pImpl->state_mutex);
    
    if (pImpl->connected) {
        return true;
    }
    
    try {
        // In production, establish actual Guardian connection
        // For now, simulate successful connection
        pImpl->connected = true;
        
        // Start event processing thread
        pImpl->event_thread_running = true;
        pImpl->event_thread = std::thread(&Impl::ProcessEventStream, pImpl.get());
        
        // Initialize metrics
        pImpl->metrics["connection_time_ms"] = 150.0;
        pImpl->metrics["guardian_count"] = 5.0;
        
        return true;
    } catch (const std::exception& e) {
        pImpl->last_error = std::make_shared<TMLError>(
            "TML_CONN_001",
            "Failed to connect to Guardian Network",
            ErrorCategory::NETWORK,
            ErrorSeverity::ERROR,
            e.what()
        );
        return false;
    }
}

bool TMLClient::ConnectWithTimeout(Duration timeout) {
    auto start = std::chrono::steady_clock::now();
    
    while (!Connect()) {
        auto elapsed = std::chrono::steady_clock::now() - start;
        if (elapsed > timeout) {
            return false;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    
    return true;
}

void TMLClient::Disconnect() {
    std::lock_guard<std::mutex> lock(pImpl->state_mutex);
    
    if (!pImpl->connected) {
        return;
    }
    
    // Flush any pending logs
    FlushLogs();
    
    // Stop event thread
    pImpl->event_thread_running = false;
    if (pImpl->event_thread.joinable()) {
        pImpl->event_thread.join();
    }
    
    pImpl->connected = false;
}

bool TMLClient::IsConnected() const {
    return pImpl->connected.load();
}

ConnectionStatus TMLClient::GetStatus() const {
    ConnectionStatus status;
    status.connected = pImpl->connected.load();
    status.guardian_url = pImpl->config->guardian_url;
    status.active_guardians = status.connected ? 5 : 0;
    status.tee_available = IsTEEAvailable();
    status.environment = pImpl->config->environment;
    status.latency = std::chrono::milliseconds(static_cast<int>(pImpl->metrics["latency_ms"]));
    status.uptime_percentage = 99.99;
    return status;
}

// ========== Sacred Zero Operations ==========

SacredZeroResult TMLClient::EvaluateSacredZero(const DecisionContext& context) {
    pImpl->stats.sacred_zero_evaluations++;
    
    // Perform Sacred Zero evaluation
    auto result = pImpl->sacred_zero->Evaluate(context);
    
    if (result.triggered) {
        pImpl->stats.sacred_zero_triggers++;
        
        // Log to Always Memory immediately
        Metadata metadata;
        metadata["violation_type"] = result.violation_type;
        metadata["must_halt"] = result.must_halt;
        metadata["trace_id"] = result.trace_id;
        
        Log(LogLevel::FATAL, "Sacred Zero Violation: " + result.description, metadata);
        
        // Trigger callbacks
        std::lock_guard<std::mutex> lock(pImpl->sacred_zero_mutex);
        for (const auto& callback : pImpl->sacred_zero_callbacks) {
            callback(result);
        }
        
        // If must halt, throw exception
        if (result.must_halt && pImpl->config->block_on_sacred_zero) {
            throw std::runtime_error("Sacred Zero Violation - System Halt Required");
        }
    }
    
    return result;
}

void TMLClient::OnSacredZeroTrigger(std::function<void(const SacredZeroResult&)> callback) {
    std::lock_guard<std::mutex> lock(pImpl->sacred_zero_mutex);
    pImpl->sacred_zero_callbacks.push_back(callback);
}

std::map<std::string, double> TMLClient::GetSacredZeroStatistics() const {
    std::map<std::string, double> stats;
    stats["total_evaluations"] = static_cast<double>(pImpl->stats.sacred_zero_evaluations);
    stats["total_triggers"] = static_cast<double>(pImpl->stats.sacred_zero_triggers);
    stats["trigger_rate"] = pImpl->stats.sacred_zero_evaluations > 0 ? 
        static_cast<double>(pImpl->stats.sacred_zero_triggers) / pImpl->stats.sacred_zero_evaluations : 0.0;
    return stats;
}

// ========== Always Memory Logging ==========

void TMLClient::Log(LogLevel level, const std::string& message, const Metadata& metadata) {
    if (!pImpl->connected && level != LogLevel::FATAL) {
        // Store locally if not connected (except for FATAL which throws)
        pImpl->always_memory->StoreLocal(level, message, metadata);
        return;
    }
    
    // Create log entry
    LogEntry entry;
    entry.level = level;
    entry.message = message;
    entry.metadata = metadata;
    entry.timestamp = std::chrono::system_clock::now();
    entry.trace_id = GenerateTraceID();
    
    // Send to Always Memory
    pImpl->always_memory->Log(entry);
    pImpl->stats.logs_sent++;
    
    // If audit mode, ensure immediate persistence
    if (pImpl->audit_mode) {
        pImpl->always_memory->Flush();
    }
}

EnvironmentalImpact TMLClient::LogEnvironmentalImpact(const ResourceUsage& resources) {
    pImpl->stats.environmental_checks++;
    
    // Calculate impact
    EnvironmentalImpact impact = CalculateEnvironmentalImpact(resources);
    
    // Update cumulative statistics
    pImpl->stats.total_carbon_kg += impact.carbon_kg;
    pImpl->stats.total_water_liters += impact.water_liters;
    
    // Log to Always Memory
    Metadata metadata;
    metadata["carbon_kg"] = impact.carbon_kg;
    metadata["water_liters"] = impact.water_liters;
    metadata["energy_kwh"] = impact.energy_kwh;
    metadata["exceeds_threshold"] = impact.exceeds_threshold;
    
    Log(impact.exceeds_threshold ? LogLevel::WARNING : LogLevel::INFO,
        "Environmental impact assessment", metadata);
    
    return impact;
}

void TMLClient::LogCompliance(ComplianceFramework framework, 
                              const std::string& status,
                              const Metadata& details) {
    Metadata metadata = details;
    metadata["framework"] = static_cast<int>(framework);
    metadata["status"] = status;
    metadata["audit_mode"] = pImpl->audit_mode.load();
    
    Log(LogLevel::INFO, "Compliance event", metadata);
}

void TMLClient::LogIndigenousData(const std::string& data_type,
                                  const std::string& community,
                                  const std::string& consent_type,
                                  const std::vector<std::string>& restrictions) {
    Metadata metadata;
    metadata["data_type"] = data_type;
    metadata["community"] = community;
    metadata["consent_type"] = consent_type;
    
    // Add restrictions as serialized string
    std::stringstream ss;
    for (const auto& r : restrictions) {
        ss << r << ";";
    }
    metadata["restrictions"] = ss.str();
    
    Log(LogLevel::INFO, "Indigenous data governance event", metadata);
}

bool TMLClient::FlushLogs() {
    if (!pImpl->connected) {
        return false;
    }
    
    try {
        return pImpl->always_memory->Flush();
    } catch (const std::exception& e) {
        pImpl->last_error = std::make_shared<TMLError>(
            "TML_FLUSH_001",
            "Failed to flush logs",
            ErrorCategory::ALWAYS_MEMORY,
            ErrorSeverity::ERROR,
            e.what()
        );
        return false;
    }
}

// ========== Environmental Monitoring ==========

EnvironmentalImpact TMLClient::CalculateEnvironmentalImpact(const ResourceUsage& resources) {
    EnvironmentalImpact impact;
    
    // Calculate various impacts
    impact.carbon_kg = pImpl->CalculateCarbonImpact(resources);
    impact.water_liters = pImpl->CalculateWaterImpact(resources);
    
    // Energy calculation
    double power_kw = (resources.cpu_cores * 0.05) + 
                     (resources.gpu_count * 0.3) +
                     (resources.memory_gb * 0.003);
    double hours = std::chrono::duration_cast<std::chrono::hours>(resources.duration).count();
    impact.energy_kwh = power_kw * hours;
    
    // E-waste estimation (simplified)
    impact.e_waste_kg = (resources.cpu_cores * 0.0001 + resources.gpu_count * 0.001) * hours / 8760;
    
    // Renewable percentage (would come from data center info)
    impact.renewable_percentage = 35.0; // Example value
    
    // Check thresholds
    impact.exceeds_threshold = CheckEnvironmentalThresholds(impact);
    
    // Generate mitigations if needed
    if (impact.exceeds_threshold) {
        impact.mitigations = SuggestMitigations(impact);
    }
    
    return impact;
}

bool TMLClient::CheckEnvironmentalThresholds(const EnvironmentalImpact& impact) {
    return impact.carbon_kg > pImpl->config->carbon_threshold_kg ||
           impact.water_liters > pImpl->config->water_threshold_liters;
}

std::vector<std::string> TMLClient::SuggestMitigations(const EnvironmentalImpact& impact) {
    std::vector<std::string> mitigations;
    
    if (impact.carbon_kg > pImpl->config->carbon_threshold_kg) {
        mitigations.push_back("Schedule workload during renewable energy peak hours");
        mitigations.push_back("Optimize algorithm efficiency to reduce compute time");
        mitigations.push_back("Consider using carbon-neutral data centers");
        mitigations.push_back("Implement model pruning to reduce resource requirements");
    }
    
    if (impact.water_liters > pImpl->config->water_threshold_liters) {
        mitigations.push_back("Use data centers with advanced cooling systems");
        mitigations.push_back("Consider edge computing to reduce centralized cooling needs");
        mitigations.push_back("Implement water recycling in cooling systems");
    }
    
    if (impact.renewable_percentage < 50) {
        mitigations.push_back("Migrate to renewable-powered data centers");
        mitigations.push_back("Purchase renewable energy certificates");
    }
    
    return mitigations;
}

// ========== Multi-stakeholder Governance ==========

std::map<std::string, double> TMLClient::EvaluateMultiStakeholder(const DecisionContext& context) {
    std::map<std::string, double> results;
    
    for (const auto& stakeholder : context.stakeholders) {
        double impact_score = 0.0;
        
        // Calculate impact based on stakeholder type
        switch (stakeholder.type) {
            case StakeholderType::HUMAN_COMMUNITY:
                // Evaluate human community impact
                impact_score = 0.3; // Simplified - would use actual impact models
                break;
                
            case StakeholderType::INDIGENOUS_COMMUNITY:
                // Indigenous communities get enhanced protection
                impact_score = 0.5;
                if (!stakeholder.consent_given) {
                    impact_score = 1.0; // Maximum negative impact without consent
                }
                break;
                
            case StakeholderType::NON_HUMAN_ENTITIES:
                // Ecosystem and biodiversity impact
                impact_score = 0.4;
                break;
                
            case StakeholderType::FUTURE_GENERATIONS:
                // Long-term sustainability impact
                impact_score = 0.35;
                break;
                
            case StakeholderType::ECOSYSTEM:
                // Overall ecosystem health
                impact_score = 0.45;
                break;
        }
        
        // Apply stakeholder weight
        impact_score *= stakeholder.weight;
        
        results[stakeholder.name] = impact_score;
        
        // Log stakeholder evaluation
        Metadata metadata;
        metadata["stakeholder_type"] = static_cast<int>(stakeholder.type);
        metadata["impact_score"] = impact_score;
        metadata["consent"] = stakeholder.consent_given;
        
        Log(LogLevel::INFO, "Stakeholder evaluation: " + stakeholder.name, metadata);
    }
    
    return results;
}

bool TMLClient::RequiresIndigenousConsent(const std::string& data_type,
                                          const std::string& location) {
    // Check if data type involves traditional knowledge
    std::vector<std::string> protected_types = {
        "traditional_knowledge",
        "sacred_sites",
        "cultural_practices",
        "medicinal_knowledge",
        "oral_history"
    };
    
    for (const auto& protected_type : protected_types) {
        if (data_type.find(protected_type) != std::string::npos) {
            return true;
        }
    }
    
    // Check if location is in indigenous territory
    // In production, this would check against a comprehensive database
    if (location.find("indigenous") != std::string::npos ||
        location.find("reservation") != std::string::npos ||
        location.find("traditional") != std::string::npos) {
        return true;
    }
    
    return pImpl->config->indigenous_data_sovereignty;
}

void TMLClient::RegisterStakeholderFeedback(StakeholderType stakeholder_type,
                                            const std::string& feedback,
                                            double impact_score) {
    Metadata metadata;
    metadata["stakeholder_type"] = static_cast<int>(stakeholder_type);
    metadata["feedback"] = feedback;
    metadata["impact_score"] = impact_score;
    
    Log(LogLevel::INFO, "Stakeholder feedback registered", metadata);
}

// ========== Compliance and Reporting ==========

ComplianceReport TMLClient::GenerateComplianceReport(Timestamp start_time,
                                                     Timestamp end_time) {
    ComplianceReport report;
    
    // Set period
    auto start = std::chrono::system_clock::to_time_t(start_time);
    auto end = std::chrono::system_clock::to_time_t(end_time);
    std::stringstream ss;
    ss << std::put_time(std::localtime(&start), "%Y-%m-%d") << " to "
       << std::put_time(std::localtime(&end), "%Y-%m-%d");
    report.period = ss.str();
    
    // Framework compliance status
    report.framework_status[ComplianceFramework::GDPR] = "COMPLIANT";
    report.framework_status[ComplianceFramework::EU_AI_ACT] = "COMPLIANT";
    report.framework_status[ComplianceFramework::UN_HUMAN_RIGHTS] = "COMPLIANT";
    report.framework_status[ComplianceFramework::PARIS_AGREEMENT] = 
        pImpl->stats.total_carbon_kg < 10000 ? "COMPLIANT" : "REVIEW_REQUIRED";
    
    // Statistics
    report.total_evaluations = pImpl->stats.sacred_zero_evaluations;
    report.violations_detected = pImpl->stats.sacred_zero_triggers;
    report.violations_prevented = pImpl->stats.sacred_zero_triggers; // All detected were prevented
    
    // Environmental impact
    report.cumulative_impact.carbon_kg = pImpl->stats.total_carbon_kg;
    report.cumulative_impact.water_liters = pImpl->stats.total_water_liters;
    
    return report;
}

std::string TMLClient::SubmitComplianceReport(const ComplianceReport& report,
                                              ComplianceFramework framework) {
    // In production, submit to regulatory API
    std::string submission_id = GenerateTraceID();
    
    Metadata metadata;
    metadata["submission_id"] = submission_id;
    metadata["framework"] = static_cast<int>(framework);
    metadata["period"] = report.period;
    
    Log(LogLevel::INFO, "Compliance report submitted", metadata);
    
    return submission_id;
}

std::map<std::string, bool> TMLClient::PerformAudit(ComplianceFramework framework) {
    std::map<std::string, bool> results;
    
    // Perform various audit checks
    results["sacred_zero_enabled"] = true;
    results["always_memory_active"] = pImpl->connected.load();
    results["environmental_monitoring"] = true;
    results["stakeholder_governance"] = true;
    results["data_sovereignty"] = pImpl->config->indigenous_data_sovereignty;
    
    return results;
}

// ========== Error Handling ==========

std::shared_ptr<TMLError> TMLClient::GetLastError() const {
    std::lock_guard<std::mutex> lock(pImpl->error_mutex);
    return pImpl->last_error;
}

void TMLClient::ClearError() {
    std::lock_guard<std::mutex> lock(pImpl->error_mutex);
    pImpl->last_error.reset();
}

void TMLClient::OnError(std::function<void(const TMLError&)> callback) {
    pImpl->error_callback = callback;
}

// ========== Configuration ==========

std::shared_ptr<Config> TMLClient::GetConfig() const {
    return pImpl->config;
}

bool TMLClient::UpdateConfig(std::shared_ptr<Config> config) {
    if (!config) {
        return false;
    }
    
    // Disconnect if connected
    bool was_connected = pImpl->connected.load();
    if (was_connected) {
        Disconnect();
    }
    
    // Update configuration
    pImpl->config = config;
    pImpl->InitializeComponents();
    
    // Reconnect if was connected
    if (was_connected) {
        return Connect();
    }
    
    return true;
}

void TMLClient::SetAuditMode(bool enable) {
    pImpl->audit_mode = enable;
    
    Metadata metadata;
    metadata["audit_mode"] = enable;
    Log(LogLevel::INFO, enable ? "Audit mode enabled" : "Audit mode disabled", metadata);
}

// ========== Monitoring and Metrics ==========

std::map<std::string, double> TMLClient::GetMetrics() const {
    std::lock_guard<std::mutex> lock(pImpl->metrics_mutex);
    auto metrics = pImpl->metrics;
    
    // Add current statistics
    metrics["sacred_zero_evaluations"] = static_cast<double>(pImpl->stats.sacred_zero_evaluations);
    metrics["sacred_zero_triggers"] = static_cast<double>(pImpl->stats.sacred_zero_triggers);
    metrics["logs_sent"] = static_cast<double>(pImpl->stats.logs_sent);
    metrics["total_carbon_kg"] = pImpl->stats.total_carbon_kg;
    metrics["total_water_liters"] = pImpl->stats.total_water_liters;
    
    return metrics;
}

std::string TMLClient::SubscribeToEvents(const std::vector<std::string>& event_types,
                                         std::function<void(const Metadata&)> callback) {
    std::lock_guard<std::mutex> lock(pImpl->subscription_mutex);
    
    std::string subscription_id = Impl::GenerateID("sub");
    
    Impl::Subscription sub;
    sub.event_types = event_types;
    sub.callback = callback;
    sub.active = true;
    
    pImpl->subscriptions[subscription_id] = sub;
    
    return subscription_id;
}

void TMLClient::UnsubscribeFromEvents(const std::string& subscription_id) {
    std::lock_guard<std::mutex> lock(pImpl->subscription_mutex);
    
    auto it = pImpl->subscriptions.find(subscription_id);
    if (it != pImpl->subscriptions.end()) {
        it->second.active = false;
        pImpl->subscriptions.erase(it);
    }
}

// ========== Utility Functions ==========

std::string TMLClient::GetVersion() {
    return "1.0.0";
}

bool TMLClient::IsTEEAvailable() {
    // Check for TEE availability
    // In production, this would check for Intel SGX, AMD SEV, or ARM TrustZone
    #ifdef __linux__
        // Check for SGX support on Linux
        return std::ifstream("/dev/sgx_enclave").good();
    #else
        return false;
    #endif
}

std::string TMLClient::GenerateTraceID() {
    return Impl::GenerateID("trace");
}

} // namespace TML

File: sdk/go/tml_client.go
/**
 * TML Client - Main interface for Ternary Moral Logic integration
 * 
 * Path: /sdk/go/tml_client.go
 * Version: 1.0.0
 * Creator: Lev Goukassian (ORCID: 0009-0006-5966-1243)
 * 
 * This is the primary interface for integrating Sacred Zero triggers
 * and Always Memory logging into Go applications.
 */

package tml

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"sync"
	"time"
)

// Decision constants for TML ternary logic
const (
	DecisionProceed    = 1  // Action approved
	DecisionSacredZero = 0  // Human review required
	DecisionRefuse     = -1 // Action blocked
)

// TMLClient is the main client for TML operations
type TMLClient struct {
	apiKey                   string
	stewardshipCouncilURL    string
	config                   *Config
	memory                   *AlwaysMemoryLogger
	sacredZero               *SacredZeroTrigger
	initialized              bool
	mu                       sync.RWMutex
	shutdownChan             chan struct{}
}

// NewClient creates a new TML client with default configuration
func NewClient(apiKey, stewardshipCouncilURL string) (*TMLClient, error) {
	return NewClientWithConfig(apiKey, stewardshipCouncilURL, DefaultConfig())
}

// NewClientWithConfig creates a new TML client with custom configuration
func NewClientWithConfig(apiKey, stewardshipCouncilURL string, config *Config) (*TMLClient, error) {
	if apiKey == "" {
		return nil, errors.New("API key is required")
	}
	if stewardshipCouncilURL == "" {
		return nil, errors.New("Stewardship Council URL is required")
	}
	if err := config.Validate(); err != nil {
		return nil, fmt.Errorf("invalid configuration: %w", err)
	}

	client := &TMLClient{
		apiKey:                apiKey,
		stewardshipCouncilURL: stewardshipCouncilURL,
		config:                config,
		shutdownChan:          make(chan struct{}),
	}

	// Initialize components
	client.memory = NewAlwaysMemoryLogger(client)
	client.sacredZero = NewSacredZeroTrigger(client)

	// Initialize client
	if err := client.initialize(); err != nil {
		return nil, fmt.Errorf("initialization failed: %w", err)
	}

	return client, nil
}

// initialize establishes connection to Stewardship Council
func (c *TMLClient) initialize() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Validate API key
	if err := c.validateAPIKey(); err != nil {
		return fmt.Errorf("API key validation failed: %w", err)
	}

	// Connect to Stewardship Council (if configured)
	if err := c.connectToStewardshipCouncil(); err != nil {
		return fmt.Errorf("Stewardship Council connection failed: %w", err)
	}

	// Verify TEE if required
	if c.config.RequireTEE {
		if err := c.verifyTEE(); err != nil {
			return fmt.Errorf("TEE verification failed: %w", err)
		}
	}

	c.initialized = true

	// Log initialization
	c.logSystemEvent("TML_CLIENT_INITIALIZED", map[string]interface{}{
		"version":              "1.0.0",
		"stewardship_council":  c.stewardshipCouncilURL,
		"tee_enabled":          c.config.RequireTEE,
	})

	return nil
}

// ProcessAction evaluates an action through TML framework
// Returns: +1 (proceed), 0 (Sacred Zero), -1 (refuse)
func (c *TMLClient) ProcessAction(ctx context.Context, action *Action) (int, error) {
	c.mu.RLock()
	if !c.initialized {
		c.mu.RUnlock()
		return DecisionSacredZero, errors.New("client not initialized")
	}
	c.mu.RUnlock()

	// Pre-action Always Memory log
	logID, err := c.memory.LogPreAction(action)
	if err != nil {
		return DecisionSacredZero, fmt.Errorf("pre-action logging failed: %w", err)
	}

	// Check Sacred Zero triggers
	decision := c.sacredZero.Evaluate(action)

	switch decision {
	case DecisionSacredZero:
		// Sacred Zero triggered - require human review
		if err := c.handleSacredZero(ctx, action, logID); err != nil {
			return decision, err
		}

	case DecisionRefuse:
		// Action refused - log and block
		if err := c.handleRefusal(ctx, action, logID); err != nil {
			return decision, err
		}
	}

	// Post-action Always Memory log
	if err := c.memory.LogPostAction(action, decision, logID); err != nil {
		return decision, fmt.Errorf("post-action logging failed: %w", err)
	}

	return decision, nil
}

// handleSacredZero processes Sacred Zero triggers
func (c *TMLClient) handleSacredZero(ctx context.Context, action *Action, logID string) error {
	event := map[string]interface{}{
		"action_id":       action.ID,
		"trigger_reason":  action.TriggerReason,
		"timestamp":       time.Now().Format(time.RFC3339),
		"requires_review": true,
	}

	// Log to Always Memory
	if err := c.memory.LogSacredZero(event, logID); err != nil {
		return err
	}

	// Notify Stewardship Council (if configured)
	c.notifyStewardshipCouncil(event)

	// If configured, wait for human review
	if c.config.BlockOnSacredZero {
		return c.waitForHumanReview(ctx, action, logID)
	}

	return nil
}

// handleRefusal processes action refusals
func (c *TMLClient) handleRefusal(ctx context.Context, action *Action, logID string) error {
	event := map[string]interface{}{
		"action_id":       action.ID,
		"refusal_reason":  action.RefusalReason,
		"timestamp":       time.Now().Format(time.RFC3339),
		"permanent":       true,
	}

	// Log to Always Memory
	if err := c.memory.LogRefusal(event, logID); err != nil {
		return err
	}

	// Notify authorities if required
	if action.RequiresLegalNotification {
		c.notifyAuthorities(event)
	}

	return nil
}

// ProcessActionAsync processes an action asynchronously
func (c *TMLClient) ProcessActionAsync(ctx context.Context, action *Action) <-chan ActionResult {
	resultChan := make(chan ActionResult, 1)

	go func() {
		decision, err := c.ProcessAction(ctx, action)
		resultChan <- ActionResult{
			Decision: decision,
			Error:    err,
		}
		close(resultChan)
	}()

	return resultChan
}

// ProcessBatch processes multiple actions in a batch
func (c *TMLClient) ProcessBatch(ctx context.Context, actions []*Action) (map[string]int, error) {
	results := make(map[string]int)
	var mu sync.Mutex

	// Create batch ID for Always Memory
	batchID := generateBatchID()
	if err := c.memory.StartBatch(batchID); err != nil {
		return nil, fmt.Errorf("batch start failed: %w", err)
	}

	// Process actions concurrently
	var wg sync.WaitGroup
	errChan := make(chan error, len(actions))

	for _, action := range actions {
		wg.Add(1)
		go func(a *Action) {
			defer wg.Done()

			decision, err := c.ProcessAction(ctx, a)
			if err != nil {
				errChan <- err
				return
			}

			mu.Lock()
			results[a.ID] = decision
			mu.Unlock()
		}(action)
	}

	wg.Wait()
	close(errChan)

	// Check for errors
	for err := range errChan {
		if err != nil {
			return results, err
		}
	}

	// Seal batch in Always Memory
	if err := c.memory.SealBatch(batchID); err != nil {
		return results, fmt.Errorf("batch seal failed: %w", err)
	}

	return results, nil
}

// GetSacredZeroStats returns Sacred Zero statistics
func (c *TMLClient) GetSacredZeroStats() *SacredZeroStats {
	return c.sacredZero.GetStatistics()
}

// GetAuditTrail retrieves audit trail for time range
func (c *TMLClient) GetAuditTrail(start, end time.Time) (*AuditTrail, error) {
	return c.memory.GetAuditTrail(start, end)
}

// validateAPIKey validates the API key with Stewardship Council
func (c *TMLClient) validateAPIKey() error {
	// Implementation would validate with Stewardship Council
	if c.apiKey == "" {
		return errors.New("invalid API key")
	}
	return nil
}

// connectToStewardshipCouncil establishes connection to Stewardship Council
func (c *TMLClient) connectToStewardshipCouncil() error {
	// Implementation would establish secure connection
	if c.stewardshipCouncilURL == "" {
		return errors.New("invalid Stewardship Council URL")
	}
	return nil
}

// verifyTEE verifies Trusted Execution Environment
func (c *TMLClient) verifyTEE() error {
	// Implementation would verify TEE attestation
	return nil
}

// waitForHumanReview waits for human review of Sacred Zero
func (c *TMLClient) waitForHumanReview(ctx context.Context, action *Action, logID string) error {
	timer := time.NewTimer(time.Duration(c.config.SacredZeroTimeout) * time.Millisecond)
	defer timer.Stop()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-timer.C:
		return errors.New("Sacred Zero timeout")
	case <-c.shutdownChan:
		return errors.New("client shutdown")
	}
}

// notifyStewardshipCouncil sends event to Stewardship Council
func (c *TMLClient) notifyStewardshipCouncil(event map[string]interface{}) {
	// Implementation would send to Stewardship Council
}

// notifyAuthorities sends notification to legal authorities
func (c *TMLClient) notifyAuthorities(event map[string]interface{}) {
	// Implementation would send to legal authorities
}

// logSystemEvent logs internal system events
func (c *TMLClient) logSystemEvent(eventType string, data map[string]interface{}) {
	c.memory.LogSystemEvent(eventType, data)
}

// Shutdown closes the client and releases resources
func (c *TMLClient) Shutdown() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if !c.initialized {
		return nil
	}

	// Signal shutdown
	close(c.shutdownChan)

	// Flush pending logs
	if err := c.memory.Flush(); err != nil {
		return fmt.Errorf("memory flush failed: %w", err)
	}

	// Disconnect from Stewardship Council
	if err := c.disconnectFromStewardshipCouncil(); err != nil {
		return fmt.Errorf("Stewardship Council disconnect failed: %w", err)
	}

	c.initialized = false
	return nil
}

// disconnectFromStewardshipCouncil closes connection to Stewardship Council
func (c *TMLClient) disconnectFromStewardshipCouncil() error {
	// Implementation would close connection cleanly
	return nil
}

// Action represents an action to be evaluated
type Action struct {
	ID                        string                 `json:"id"`
	Type                      string                 `json:"type"`
	Content                   string                 `json:"content"`
	Actor                     string                 `json:"actor"`
	Target                    string                 `json:"target"`
	Context                   map[string]interface{} `json:"context"`
	TriggerReason             string                 `json:"trigger_reason,omitempty"`
	RefusalReason             string                 `json:"refusal_reason,omitempty"`
	RequiresLegalNotification bool                   `json:"requires_legal_notification"`
}

// ActionResult represents the result of an async action processing
type ActionResult struct {
	Decision int
	Error    error
}

// Helper functions

func generateBatchID() string {
	timestamp := time.Now().UnixNano()
	hash := sha256.Sum256([]byte(fmt.Sprintf("batch-%d", timestamp)))
	return hex.EncodeToString(hash[:])[:16]
}
